### ISSUE TITLE
TcpKeepAliveSettings doesn't work on Linux

### ISSUE THREAD

**Issue by Havret** [Azure/amqpnetlite#398]
I am getting the following exception when, I try to use `TcpKeepAliveSettings`  with the app deployed on Linux box. I'm not sure but this should be supported https://github.com/dotnet/corefx/pull/29963. Do we need to make some adjustments in AmqpNetLite to make this work properly on Linux? 

```
Unhandled exception. Apache.NMS.NMSException: Socket.IOControl handles Windows-specific control codes and is not supported on this platform.
---> System.PlatformNotSupportedException: Socket.IOControl handles Windows-specific control codes and is not supported on this platform.
   at System.Net.Sockets.SocketPal.WindowsIoctl(SafeSocketHandle handle, Int32 ioControlCode, Byte[] optionInValue, Byte[] optionOutValue, Int32& optionLength)
   at System.Net.Sockets.Socket.IOControl(Int32 ioControlCode, Byte[] optionInValue, Byte[] optionOutValue)
   at System.Net.Sockets.Socket.IOControl(IOControlCode ioControlCode, Byte[] optionInValue, Byte[] optionOutValue)
   at Amqp.SocketExtensions.SetTcpKeepAlive(Socket socket, TcpKeepAliveSettings settings)
   at Amqp.TcpSettings.Configure(Socket socket)
   at Amqp.TcpTransport.ConnectAsync(Address address, ConnectionFactory factory)
   at Amqp.ConnectionFactory.CreateAsync(Address address, Open open, OnOpened onOpened, IHandler handler)
   at Apache.NMS.AMQP.Provider.Amqp.AmqpConnection.Start()
   at Apache.NMS.AMQP.NmsConnection.CreateNmsConnection()
   --- End of inner exception stack trace ---
   at Apache.NMS.AMQP.NmsConnection.CreateNmsConnection()
   at Apache.NMS.AMQP.NmsConnection.CreateSession(AcknowledgementMode acknowledgementMode)
```

**Comment by xinchen10** [comment]
netstandard1.3 or 2.0? 1.3 is still using very old 4.1.0 System.Net.Sockets package released in 2016. The fix was made in 2018. Have you tried 2.0 with the latest .net core runtime?

**Comment by HavretGC** [comment]
I think I'm using the new version. Here you can have a look at the app I sent to Red Hat to help them reproduce the issue --> https://github.com/HavretGC/TcpKeepAliveSettings

**Comment by HavretGC** [comment]
Can't we address this the same way as it was done for SqlClient? --> https://github.com/dotnet/SqlClient/pull/395

**Comment by xinchen10** [comment]
That requires us to build multiple targets and lose the benefit of targeting just netstandard. This is what I have found.
* the new socket options were added in this PR https://github.com/dotnet/corefx/pull/29963, which should be available since .Net Core 3.0 releases.
* however the new options are not available in netstandard2.0. The SetSocketOption API is available and it seems that it simply passes the values directly to the PAL implementation. So technically we can send the new values (just not the Enum).

Branch `socketopt` has the changes based on the above assumption. Can you try it on .Net Core 3.0+ runtime on Linux?

**Comment by Havret** [comment]
I was able to run my test app with the changes from `socketopt` branch on NET Core 3.1 runtime on Linux box. It seems to work, or at least it doesn't throw the exception anymore. I'm not sure how can I make sure that TcpKeepAliveSettings are applied and work as expected though. Probably that would require some packet sniffing with Wireshark or sth like that. 

Possibly we could ask guys from Red Hat, to confirm that the fix is ok.

**Comment by Havret** [comment]
@xinchen10 as per Red Hat:

> I did some more digging and review of NET core runtime code and the code given in the patch seems like it ought to work in cases where the platform supports those options.  The code would probably still fail with a platform exception if the OS in question doesn't such as older versions of Windows 10 or a linux distro where the TCP specific options they are using from "/usr/include/netinet/tcp.h" which is what the net core uses via a CMake check to determine if those options are present.  

> Is that helpful ? If not what more information you need. Being specific to environment, it wouldn't be official fix.

So do we need anything more carry on with that?

**Comment by michaelpearce-gain** [comment]
LGTM to be honest.

**Comment by xinchen10** [comment]
I have merged the change to master.

**Comment by HavretGC** [comment]
Can we expect the release any time soon? :)

**Comment by xinchen10** [comment]
Released 2.4.0.
